src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -799,7 +799,8 @@ public E get(int i) {
                 }
                 final E newValue = mapper.apply(i);
                 if (!stable.trySet(newValue)) {
-                    throw new IllegalStateException(
+                    // This should never happen
+                    throw new InternalError(
```

Comment:
This can happen when the mapper recursively calls the `get` method on this list.
```suggestion
                    throw new ConcurrentModificationException(
```
or
```suggestion
                    throw new IllegalStateException(
```

Reply:
I think, if the mapper is recursive, we never reach the `stable.trySet()` statement. I've added a test to make sure this is the case:

```
    @Test
    void recursiveCall() {
        AtomicReference<IntFunction<Integer>> ref = new AtomicReference<>();
        var lazy = StableValue.lazyList(SIZE, i -> ref.get().apply(i));
        ref.set(lazy::get);
        assertThrows(StackOverflowError.class, () -> lazy.get(INDEX));
    }
```

Is there another way we could end up there?

Reply:
The following makes it happen:
```java
var mapper = new AtomicReference<IntFunction<Integer>>();
var list = StableValue.lazyList(SIZE, i -> mapper.get().apply(i));

var recursing = new boolean[SIZE];
mapper.set(i -> {
	if (recursing[i]) {
		return i;
	}

	recursing[i] = true;
	return list.get(i);
});

list.get(INDEX);
```

Reply:
Ahh. But that is cheating.... ;-)  Well, I see what you mean now. I will update the exception type.